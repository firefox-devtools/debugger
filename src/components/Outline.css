.outline-list {
  list-style-type: "-";
}

.outline-list__element {
  color: blue;
  padding-left: 0.5rem;
}
