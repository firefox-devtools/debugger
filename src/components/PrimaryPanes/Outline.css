.outline {
  overflow-y: auto;
}

.outline .outline-pane-info {
  width: 100%;
  font-style: italic;
  text-align: center;
  padding: 0.5em;
  user-select: none;
  font-size: 12px;
}

.outline-list {
  list-style-type: none;
  padding-left: 0px;
  width: 100%;
}

.outline-list__element {
  color: blue;
  padding-bottom: 0.2rem;
  padding-left: 1rem;
  padding-right: 0.5rem;
  padding-top: 0.2rem;
  cursor: default;
}

.outline-list__element:hover {
  background: var(--theme-toolbar-background-hover);
}
