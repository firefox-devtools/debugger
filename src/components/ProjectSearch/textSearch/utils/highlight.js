// Maybe reuse file search's functions?
import React from "react";

export function highlightMatches(lineMatch) {
  const { value, column, match } = lineMatch;
  const len = match.length;

  return (
    <span className="line-value">
      <span className="line-match" key={0}>
        {value.slice(0, column)}
      </span>
      <span className="query-match" key={1}>
        {value.substr(column, len)}
      </span>
      <span className="line-match" key={2}>
        {value.slice(column + len, value.length)}
      </span>
    </span>
  );
}
