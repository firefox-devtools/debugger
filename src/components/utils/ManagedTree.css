.tree {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  -o-user-select: none;
  user-select: none;

  flex: 1;
  white-space: nowrap;
  overflow: auto;
}

.tree button {
  display: block;
}

.tree .node {
  padding: 2px 5px;
  position: relative;
}

.tree .node.focused {
  color: white;
  background-color: var(--theme-selection-background);
}

html:not([dir="rtl"]) .tree .node > div {
  margin-left: 10px;
}

html[dir="rtl"] .tree .node > div {
  margin-right: 10px;
}

.tree .node.focused svg {
  fill: white;
}

.tree-node button {
  position: fixed;
}
