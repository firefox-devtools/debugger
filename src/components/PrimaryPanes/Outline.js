// @flow

import React, { Component } from "react";
import { bindActionCreators } from "redux";
import { connect } from "react-redux";
import classnames from "classnames";
import actions from "../../actions";
import { getSelectedSource, getSymbols } from "../../selectors";
import { isEnabled } from "devtools-config";
import "./Outline.css";
import previewFunction from "../shared/previewFunction";

import type {
  SymbolDeclarations,
  SymbolDeclaration
} from "../../utils/parser/getSymbols";
import type { SourceRecord } from "../../reducers/sources";

type State = any;
type Props = {
  isHidden: boolean,
  symbols: SymbolDeclarations,
  selectSource: (string, { line: number }) => void,
  selectedSource: ?SourceRecord
};

class Outline extends Component<Props, State> {
  constructor(props) {
    super(props);
    this.state = {};
  }

  selectItem(location) {
    const { selectedSource, selectSource } = this.props;
    if (!selectedSource) {
      return;
    }
    const selectedSourceId = selectedSource.get("id");
    const startLine = location.start.line;
    selectSource(selectedSourceId, { line: startLine });
  }

  renderFunction(func: SymbolDeclaration) {
    const { name, location } = func;

    return (
      <li
        key={`${name}:${location.start.line}:${location.start.column}`}
        className="outline-list__element"
        onClick={() => this.selectItem(location)}
      >
        {previewFunction({ name })}
      </li>
    );
  }

  renderFunctions() {
    const { symbols } = this.props;

    return symbols.functions
      .filter(func => func.name != "anonymous")
      .map(func => this.renderFunction(func));
  }

  render() {
    const { isHidden } = this.props;
    if (!isEnabled("outline")) {
      return null;
    }

    return (
      <div className={classnames("outline", { hidden: isHidden })}>
        <ul className="outline-list">
          {this.renderFunctions()}
        </ul>
      </div>
    );
  }
}

Outline.displayName = "Outline";

export default connect(
  state => {
    const selectedSource = getSelectedSource(state);
    return {
      symbols: getSymbols(state, selectedSource && selectedSource.toJS()),
      selectedSource
    };
  },
  dispatch => bindActionCreators(actions, dispatch)
)(Outline);
