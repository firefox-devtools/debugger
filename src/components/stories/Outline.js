/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import PropTypes from "prop-types";
import React from "react";
import { storiesOf } from "@storybook/react";
import { action } from "@storybook/addon-actions";

import { Outline } from "../PrimaryPanes/Outline";
import { L10N } from "devtools-launchpad";

import "../App.css";
import "../PrimaryPanes/Outline.css";

require("devtools-modules/src/themes/dark-theme.css");

if (typeof window == "object") {
  window.L10N = L10N;
  window.L10N.setBundle(require("../../../assets/panel/debugger.properties"));
}

function makeFuncLocation(startLine) {
  return {
    start: {
      line: startLine
    }
  };
}

function makeSymbolDeclaration(name, line, klass = "") {
  return {
    id: `${name}:${line}`,
    name,
    location: makeFuncLocation(line),
    klass
  };
}

function OutlineFactory({ dir = "ltr", theme = "dark", ...props }) {
  const themeClass = `theme-${theme}`;
  document.body.parentNode.className = themeClass;
  return (
    <div
      className={`outline ${themeClass}`}
      dir={dir}
      style={{
        width: "300px",
        margin: "40px",
        border: "1px solid var(--theme-splitter-color)"
      }}
    >
      <Outline
        selectLocation={action("selectFrame")}
        isHidden={false}
        {...props}
      />
    </div>
  );
}

OutlineFactory.propTypes = {
  dir: PropTypes.string,
  theme: PropTypes.string
};

storiesOf("Outline", module)
  .add("empty", () => {
    const symbols = { functions: [], variables: [] };
    return <OutlineFactory symbols={symbols} />;
  })
  .add("simple", () => {
    const symbols = {
      functions: [
        makeSymbolDeclaration("foo", 2),
        makeSymbolDeclaration("render", 2)
      ],
      variables: []
    };
    return <OutlineFactory symbols={symbols} />;
  })
  .add("component", () => {
    const symbols = {
      functions: [
        makeSymbolDeclaration("foo", 2, "Fancy"),
        makeSymbolDeclaration("render", 2, "Fancy")
      ],
      variables: []
    };
    return <OutlineFactory symbols={symbols} />;
  })
  .add("complex component", () => {
    const symbols = {
      functions: [
        makeSymbolDeclaration("foo", 2, "Fancy"),
        makeSymbolDeclaration("render", 2, "Fancy"),
        makeSymbolDeclaration("other", 2),
        makeSymbolDeclaration("utilBar", 2)
      ],
      variables: []
    };
    return <OutlineFactory symbols={symbols} />;
  });
