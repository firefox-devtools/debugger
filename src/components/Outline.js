// @flow

import { DOM as dom, PropTypes, Component } from "react";
import { bindActionCreators } from "redux";
import { connect } from "react-redux";
import actions from "../actions";
import { getSelectedSource, getSourceText } from "../selectors";
import { isEnabled } from "devtools-config";
import { getSymbols } from "../utils/parser";
import "./Outline.css";

import type { Record } from "../utils/makeRecord";
import type { SourceText } from "debugger-html";

class Outline extends Component {
  state: any;

  constructor(props) {
    super(props);
    this.state = {};
  }

  componentWillReceiveProps({ sourceText }) {
    if (sourceText) {
      this.setSymbolDeclarations(sourceText);
    }
  }

  async setSymbolDeclarations(sourceText: Record<SourceText>) {
    const symbolDeclarations = await getSymbols(sourceText.toJS());

    if (symbolDeclarations !== this.state.symbolDeclarations) {
      this.setState({
        symbolDeclarations
      });
    }
  }

  renderFunction(func) {
    return dom.li({ key: func.id }, func.value);
  }

  renderFunctions() {
    const { symbolDeclarations } = this.state;
    if (!symbolDeclarations) {
      return;
    }

    const { functions } = symbolDeclarations;

    return functions
      .filter(func => func.value != "anonymous")
      .map(this.renderFunction);
  }

  render() {
    if (!isEnabled("outline")) {
      return null;
    }

    return dom.div(
      { className: "outline" },
      dom.ul({}, this.renderFunctions())
    );
  }
}

Outline.propTypes = {
  selectedSource: PropTypes.object
};

Outline.displayName = "Outline";

export default connect(
  state => {
    const selectedSource = getSelectedSource(state);
    const sourceId = selectedSource ? selectedSource.get("id") : null;

    return {
      sourceText: getSourceText(state, sourceId)
    };
  },
  dispatch => bindActionCreators(actions, dispatch)
)(Outline);
