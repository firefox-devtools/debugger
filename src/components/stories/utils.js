const renderComponentFromFixture =
      require("../../test/utils/renderComponentFromFixture");

const { storiesOf } = require("@kadira/storybook");

module.exports = {
  renderComponent: renderComponentFromFixture,
  storiesOf
};
