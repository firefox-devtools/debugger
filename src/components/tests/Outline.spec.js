/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import React from "react";
import { shallow } from "enzyme";
import Outline from "../../components/PrimaryPanes/Outline";
import devtoolsConfig from "devtools-config";
import { makeSymbolDeclaration } from "../../utils/test-head";
import { showMenu } from "devtools-contextmenu";
import { copyToTheClipboard } from "../../utils/clipboard";

import {
  getShownSource,
  getSelectedSource,
  getDebuggeeUrl,
  getExpandedState,
  getProjectDirectoryRoot,
  getSources
} from "../../selectors";

jest.mock("devtools-contextmenu", () => ({ showMenu: jest.fn() }));
jest.mock("../../utils/clipboard", () => ({ copyToTheClipboard: jest.fn() }));
jest.mock("../../selectors");


const sourceId = "id";
const mockFunctionText = "mock function text";

function generateDefaults(overrides) {
  return {
    selectLocation: jest.genMockFunction(),
    selectedSource: {
      get: () => sourceId
    },
    getFunctionText: jest.fn().mockReturnValue(mockFunctionText),
    flashLineRange: jest.fn(),
    isHidden: false,
    symbols: {},
    selectedLocation: {
      sourceId: sourceId
    },
    ...overrides
  };
}

function render(overrides = {}) {
  const props = generateDefaults(overrides);
  const component = shallow(<Outline.WrappedComponent {...props} />);
  const instance = component.instance();
  return { component, props, instance };
}

describe("Outline", () => {
  beforeEach(() => {
    devtoolsConfig.isEnabled = jest.fn();
    devtoolsConfig.isEnabled.mockReturnValue(true);
  });

  afterEach(() => {
    copyToTheClipboard.mockClear();
    showMenu.mockClear();
  });

  it("should render a list of functions when properties change", async () => {
    const symbols = {
      functions: [
        makeSymbolDeclaration("my_example_function1", 21),
        makeSymbolDeclaration("my_example_function2", 22)
      ]
    };

    const { component } = render({ symbols });
    expect(component).toMatchSnapshot();
  });


  it("should select a line of code in the current file on click", async () => {
    const startLine = 12;
    const symbols = {
      functions: [makeSymbolDeclaration("my_example_function", startLine)]
    };

    const { component, props } = render({ symbols });

    const { selectLocation } = props;
    const listItem = component.find("li").first();
    listItem.simulate("click");
    expect(selectLocation).toHaveBeenCalledWith({ line: startLine, sourceId });
  });

  describe("renders outline", () => {
    it("renders placeholder `No File Selected` if no selectedSource is defined", () => {
      const { component } = render({
        selectedSource: null
      });
      expect(component).toMatchSnapshot();
    });
    it("renders loading if symbols is not defined", () => {
      const { component } = render({
        symbols: null
      });
      expect(component).toMatchSnapshot();
    });
    it("renders loading if symbols are loading", () => {
      const { component } = render({
        symbols: {
          loading: true
        }
      });
      expect(component).toMatchSnapshot();
    });
    it("should render ignore anonymous functions", async () => {
      const symbols = {
        functions: [
          makeSymbolDeclaration("my_example_function1", 21),
          makeSymbolDeclaration("anonymous", 25)
        ]
      };

      const { component } = render({ symbols });
      expect(component).toMatchSnapshot();
    });
    it("should render placholder `No functions` if all func are anonymous", async () => {
      const symbols = {
        functions: [
          makeSymbolDeclaration("anonymous", 25),
          makeSymbolDeclaration("anonymous", 30)
        ]
      };

      const { component } = render({ symbols });
      expect(component).toMatchSnapshot();
    });
    it("should render placholder `No functions` if symbols has no functions", async () => {
      const symbols = {
        functions: [
        ]
      };
      const { component } = render({ symbols });
      expect(component).toMatchSnapshot();
    });
    it("should sort functions alphabetically by function name", async () => {
      const symbols = {
        functions: [
          makeSymbolDeclaration("c_function", 25),
          makeSymbolDeclaration("x_function", 30),
          makeSymbolDeclaration("a_function", 70)
        ]
      };

      const { component } = render({ symbols:symbols,  alphabetizeOutline: true});
      expect(component).toMatchSnapshot();
    });
    it("should render functions by function class", async () => {
      const symbols = {
        functions: [
          makeSymbolDeclaration("x_function", 25, 26, "x_klass"),
          makeSymbolDeclaration("a2_function", 30, 31, "a_klass"),
          makeSymbolDeclaration("a1_function", 70, 71, "a_klass")
        ],
        classes: [
          makeSymbolDeclaration("x_klass", 24, 27),
          makeSymbolDeclaration("a_klass", 29, 72)
        ]
      };

      const { component } = render({ symbols:symbols });
      expect(component).toMatchSnapshot();
    });
    it("should render functions by function class, alphabetically", async () => {
      const symbols = {
        functions: [
          makeSymbolDeclaration("x_function", 25, 26, "x_klass"),
          makeSymbolDeclaration("a2_function", 30, 31, "a_klass"),
          makeSymbolDeclaration("a1_function", 70, 71, "a_klass")
        ],
        classes: [
          makeSymbolDeclaration("x_klass", 24, 27),
          makeSymbolDeclaration("a_klass", 29, 72)
        ]
      };

      const { component } = render({ symbols:symbols,  alphabetizeOutline: true });
      expect(component).toMatchSnapshot();
    });
  });

  describe("onContextMenu of Outline", () => {
    it("is called onContextMenu for each item", async () => {
      const event = {event: "oncontextmenu"};
      const fn = makeSymbolDeclaration("exmple_function", 2);
      const symbols = {
        functions: [
          fn,
        ]
      };

      const { component, instance } = render({ symbols });
      instance.onContextMenu = jest.fn(() => {});
      await component
            .find(".outline-list__element")
            .simulate("contextmenu", event);

      expect(instance.onContextMenu).toHaveBeenCalledWith(event, fn);

    });
    it("does not show menu with no selected source", async () => {
      const mockEvent = {
        preventDefault: jest.fn(),
        stopPropagation: jest.fn()
      };
      const { instance } = render({
        selectedSource: null
      });
      await instance.onContextMenu(mockEvent, {});
      expect(mockEvent.preventDefault).toHaveBeenCalled();
      expect(mockEvent.stopPropagation).toHaveBeenCalled();
      expect(showMenu).not.toHaveBeenCalled();
    });

    it("shows menu to copy function on context menu, and calls copy function on click", async () => {
      const startLine = 12;
      const endLine = 21;
      const func = makeSymbolDeclaration("my_example_function", startLine, endLine);
      const symbols = {
        functions: [func]
      };
      const mockEvent = {
        preventDefault: jest.fn(),
        stopPropagation: jest.fn()
      };
      const { instance, props } = render({ symbols });
      await instance.onContextMenu(mockEvent, func);

      expect(mockEvent.preventDefault).toHaveBeenCalled();
      expect(mockEvent.stopPropagation).toHaveBeenCalled();

      const expectedMenuOptions = [
        {
          accesskey: "F",
          click: expect.any(Function),
          disabled: false,
          id: "node-menu-copy-function",
          label: "Copy function"
        }
      ];
      expect(props.getFunctionText).toHaveBeenCalledWith(12);
      expect(showMenu).toHaveBeenCalledWith(mockEvent, expectedMenuOptions);

      showMenu.mock.calls[0][1][0].click();
      expect(copyToTheClipboard).toHaveBeenCalledWith(mockFunctionText);
      expect(props.flashLineRange).toHaveBeenCalledWith({"end": endLine, "sourceId": sourceId, "start": startLine});
    });

    describe("test redux connect", () => {
      jest.mock("../../selectors");
      it("calls mapStateToProps", async () => {
        const state = { hello: "world" };
        const store = {
          getState: () => state,
          dispatch: () => {},
          subscribe: () => {}
        };
        shallow(<Outline store={store} />);
      });
    });
  });
});
