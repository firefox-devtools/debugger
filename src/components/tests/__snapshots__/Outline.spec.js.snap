// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Outline should render a list of functions when properties change 1`] = `
<div
  className="outline"
>
  <ul
    className="outline-list"
  >
    <li
      className="outline-list__element"
      key="my_example_function1:21:undefined"
      onClick={[Function]}
    >
      <PreviewFunction
        func={
          Object {
            "name": "my_example_function1",
            "parameterNames": undefined,
          }
        }
      />
    </li>
    <li
      className="outline-list__element"
      key="my_example_function2:22:undefined"
      onClick={[Function]}
    >
      <PreviewFunction
        func={
          Object {
            "name": "my_example_function2",
            "parameterNames": undefined,
          }
        }
      />
    </li>
  </ul>
</div>
`;

exports[`Outline should render ignore anonimous functions 1`] = `
<div
  className="outline"
>
  <ul
    className="outline-list"
  >
    <li
      className="outline-list__element"
      key="my_example_function1:21:undefined"
      onClick={[Function]}
    >
      <PreviewFunction
        func={
          Object {
            "name": "my_example_function1",
            "parameterNames": undefined,
          }
        }
      />
    </li>
  </ul>
</div>
`;
